confmap/provider/googlesecretmanagerprovider/provider.go

```diff
@@ -0,0 +1,86 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//go:generate mdatagen metadata.yaml
+
+package googlesecretmanagerprovider // import "github.com/open-telemetry/opentelemetry-collector-contrib/confmap/provider/googlesecretmanagerprovider"
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	gax "github.com/googleapis/gax-go/v2"
+	"github.com/googleapis/gax-go/v2/apierror"
+	"go.opentelemetry.io/collector/confmap"
+
+	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
+)
+
+type secretsManagerClient interface {
+	AccessSecretVersion(ctx context.Context, req *secretmanagerpb.AccessSecretVersionRequest, opts ...gax.CallOption) (*secretmanagerpb.AccessSecretVersionResponse, error)
+	Close() error
+}
+
+const (
+	schemeName = "googlesecretmanagerprovider"
+)
+
+type provider struct {
+	mu     sync.Mutex
+	client secretsManagerClient
+}
+
+func NewFactory() confmap.ProviderFactory {
+	return confmap.NewProviderFactory(newProvider)
+}
+
+func newProvider(confmap.ProviderSettings) confmap.Provider {
+	return &provider{client: nil}
+}
+
+func (p *provider) Retrieve(ctx context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
+	if !strings.HasPrefix(uri, schemeName+":") {
+		return nil, fmt.Errorf("%q uri is not supported by Google Secret Manager Provider", uri)
+	}
+	secretName := strings.TrimPrefix(uri, schemeName+":")
+	p.mu.Lock()
+	if p.client == nil {
```

Comment:
I don't think I see the upside in the client not being instantiated on initialization. This would eliminate the need for a mutex.

Reply:
HTTP Provider initializes http client from within `Retrieve()`: https://github.com/open-telemetry/opentelemetry-collector/blob/c85ebbec779fa00c3fbe4a555eb0d420bdad7581/confmap/provider/internal/configurablehttpprovider/provider.go#L89
So as the AWS secret manager Provider:    https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/54eb1c1d699ed630432b47ebe3f70580dc8dd459/confmap/provider/secretsmanagerprovider/provider.go#L56-L61

Also, Placing the secret manager client initialization call within `NewFactory()` would prevent errors from being surfaced. If the Secret Manager initialization fails, there would be no way to communicate that failure back to the caller if it's triggered within `NewFactory()`.